arwen/elrondapi/managedei.go

```diff
@@ -0,0 +1,1035 @@
+	managedType := arwen.GetManagedTypesContext(context)
+	runtime := arwen.GetRuntimeContext(context)
+	metering := arwen.GetMeteringContext(context)
+
+	gasToUse := metering.GasSchedule().ElrondAPICost.SignalError
+	metering.UseGas(gasToUse)
+
+	errBytes, err := managedType.GetBytes(errHandle)
+	if arwen.WithFault(err, context, runtime.ManagedBufferAPIErrorShouldFailExecution()) {
+		return
+	}
+	managedType.ConsumeGasForThisIntNumberOfBytes(len(errBytes))
+
+	runtime.SignalUserError(string(errBytes))
+}
+
+//export v1_4_managedWriteLog
+func v1_4_managedWriteLog(
+	context unsafe.Pointer,
+	topicsHandle int32,
+	dataHandle int32,
+) {
+	runtime := arwen.GetRuntimeContext(context)
+	output := arwen.GetOutputContext(context)
+	metering := arwen.GetMeteringContext(context)
+	managedType := arwen.GetManagedTypesContext(context)
+
+	topics, sumOfTopicByteLengths, err := readManagedVecOfManagedBuffers(managedType, topicsHandle)
+	if arwen.WithFault(err, context, runtime.ElrondAPIErrorShouldFailExecution()) {
+		return
+	}
+
+	dataBytes, err := managedType.GetBytes(dataHandle)
+	if arwen.WithFault(err, context, runtime.ManagedBufferAPIErrorShouldFailExecution()) {
+		return
+	}
+	managedType.ConsumeGasForThisIntNumberOfBytes(len(dataBytes))
+	dataByteLen := uint64(len(dataBytes))
+
+	gasToUse := metering.GasSchedule().ElrondAPICost.Log
+	gasForData := math.MulUint64(
+		metering.GasSchedule().BaseOperationCost.DataCopyPerByte,
+		sumOfTopicByteLengths+dataByteLen)
+	gasToUse = math.AddUint64(gasToUse, gasForData)
+	metering.UseGas(gasToUse)
+
+	output.WriteLog(runtime.GetSCAddress(), topics, dataBytes)
+}
+
+//export v1_4_managedGetOriginalTxHash
+func v1_4_managedGetOriginalTxHash(context unsafe.Pointer, resultHandle int32) {
+	runtime := arwen.GetRuntimeContext(context)
+	metering := arwen.GetMeteringContext(context)
+	managedType := arwen.GetManagedTypesContext(context)
+
+	gasToUse := metering.GasSchedule().ElrondAPICost.GetBlockHash
+	metering.UseGas(gasToUse)
+
+	managedType.SetBytes(resultHandle, runtime.GetOriginalTxHash())
+}
+
+//export v1_4_managedGetStateRootHash
+func v1_4_managedGetStateRootHash(context unsafe.Pointer, resultHandle int32) {
+	blockchain := arwen.GetBlockchainContext(context)
+	metering := arwen.GetMeteringContext(context)
+	managedType := arwen.GetManagedTypesContext(context)
+
+	gasToUse := metering.GasSchedule().ElrondAPICost.GetStateRootHash
+	metering.UseGas(gasToUse)
+
+	managedType.SetBytes(resultHandle, blockchain.GetStateRootHash())
+}
+
+//export v1_4_managedGetBlockRandomSeed
+func v1_4_managedGetBlockRandomSeed(context unsafe.Pointer, resultHandle int32) {
+	blockchain := arwen.GetBlockchainContext(context)
+	metering := arwen.GetMeteringContext(context)
+	managedType := arwen.GetManagedTypesContext(context)
+
+	gasToUse := metering.GasSchedule().ElrondAPICost.GetBlockRandomSeed
+	metering.UseGas(gasToUse)
+
+	managedType.SetBytes(resultHandle, blockchain.CurrentRandomSeed())
```

Comment:
Catastrophically, `blockchain.CurrentRandomSeed()` returns the slice that has been set as the current random seed in the block header. The actual slice instance. Then the contract can modify it.